climada/util/lines_polys_handler.py

```diff
@@ -0,0 +1,1059 @@
+
+    return pnt_mask, line_mask, poly_mask
+
+
+def gdf_to_pnts(gdf, res, to_meters):
+    """
+    Disaggregate geodataframe with (multi-)polygons geometries to points.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        Feodataframe instance with gdf.geometry containing (multi)-lines or (multi-)polygons.
+        Points are ignored.
+    res : float
+        Resolution of the disaggregation grid. Can also be a tuple of [x_grid, y_grid] numpy
+        arrays. In this case, to_meters is ignored.
+    to_meters : bool
+       If True, the geometries are projected to an equal area projection before the disaggregation.
+       res is then in meters. The exposures are then reprojected into the original projections
+       before the impact calculation.
+
+    Returns
+    -------
+    gdf_pnt : gpd.GeoDataFrame
+        with a double index, first for the geometries of exp, second for the point disaggregation
+        of the geometries.
+
+    """
+    if gdf.empty:
+        return gdf
+
+    pnt_mask, line_mask, poly_mask = _pnt_line_poly_mask(gdf)
+
+    # Concatenating an empty dataframe with an index together with
+    # a dataframe with a multi-index breaks the multi-index
+    gdf_pnt = gpd.GeoDataFrame([])
+    if pnt_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            gdf[pnt_mask]
+        ]))
+    if line_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            _line_to_pnts(gdf[line_mask], res, to_meters)
+        ]))
+    if poly_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            _poly_to_pnts(gdf[poly_mask], res, to_meters)
+        ]))
+
+    return gdf_pnt
+
+
+def gdf_to_grid(gdf, grid):
+    """
+    Disaggregate geodataframe with (multi-)polygons geometries to points.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        Feodataframe instance with gdf.geometry containing (multi)-lines or (multi-)polygons.
+        Points are ignored.
+    grid : np.array()
+        Grid on which to disaggregate the exposures. Provided as two vectors [x_grid, y_grid].
+
+    Returns
+    -------
+    gdf_pnt : gpd.GeoDataFrame
+        with a double index, first for the geometries of exp, second for the point disaggregation
+        of the geometries.
+    """
+    if gdf.empty:
+        return gdf
+
+    pnt_mask, line_mask, poly_mask = _pnt_line_poly_mask(gdf)
+
+    # Concatenating an empty dataframe with an index together with
+    # a dataframe with a multi-index breaks the multi-index
+    gdf_pnt = gpd.GeoDataFrame([])
+    if pnt_mask.any():
+        gdf_pnt = gpd.GeoDataFrame(pd.concat([
+            gdf_pnt,
+            gdf[pnt_mask]
+        ]))
+    if line_mask.any():
+        raise AttributeError("The dataframe contains lines. Lines cannot be disaggregated onto a "
+                             "fixed grid.")
```

Comment:
Why is it that points can be disaggregated onto a fixed grid but lines cannot?

Reply:
allowing to keep points in a mixed point / poly dataframe results in having a disaggregated dataframe in the end that has single indices (in the point section) and multi-indices (in the poly section). This then cause issues in the re-aggregation, which expects strictly a multi-index.

Hence I suggest to either:
- allow disaggregation on _any_ mixed line / point / poly dataframe. That would need a method to temporarily set a multi-index on original point geometries and a method to disaggregate a line onto a grid, too.
- allow disaggregation to grid only for polygon gdf

I advocate for now for the latter, and keep the first as an extension option in the issues section.